src/scf/self_consistent_field.jl

```diff
@@ -133,25 +91,156 @@ Solve the Kohn-Sham equations with a SCF algorithm, starting at ρ.
     callback(info)
     info
 end
+
+
+"""
+Obtain new density ρ by diagonalizing `ham`. Follows the policy imposed by the `bands`
+data structure to determine and adjust the number of bands to be computed.
+"""
+function next_density(ham::Hamiltonian, bands=AdaptiveBands(ham.basis.model);
+                      eigensolver=lobpcg_hyper, ψ=nothing, eigenvalues=nothing,
+                      occupation=nothing, kwargs...)
+    n_bands_converge, n_bands_compute = determine_n_bands(bands, occupation, eigenvalues, ψ)
+
+    if isnothing(ψ)
+        increased_n_bands = true
+    else
+        @assert length(ψ) == length(ham.basis.kpoints)
+        n_bands_compute = max(n_bands_compute, maximum(ψk -> size(ψk, 2), ψ))
+        increased_n_bands = n_bands_compute > size(ψ[1], 2)
+    end
+
+    # TODO Synchronize since right now it is assumed that the same number of bands are
+    #      computed for each k-Point
+    n_bands_compute = mpi_max(n_bands_compute, ham.basis.comm_kpts)
+
+    eigres = diagonalize_all_kblocks(eigensolver, ham, n_bands_compute; ψguess=ψ,
+                                     n_conv_check=n_bands_converge, kwargs...)
+    eigres.converged || (@warn "Eigensolver not converged" iterations=eigres.iterations)
+
+    # Check maximal occupation of the unconverged bands is sensible.
+    occupation, εF = compute_occupation(ham.basis, eigres.λ)
+    minocc = maximum(minimum, occupation)
+
+    # TODO This is a bit hackish, but needed right now as we increase the number of bands
+    #      to be computed only between SCF steps. Should be revisited once we have a better
+    #      way to deal with such things in LOBPCG.
+    if !increased_n_bands && minocc > bands.occupation_threshold
+        @warn("Detected large minimal occupation $minocc. SCF could be unstable. " *
+              "Try switching to adaptive band selection (`bands=AdaptiveBands(basis)`) " *
+              "or request more converged bands than $n_bands_converge (e.g. " *
+              "`bands=AdaptiveBands(basis; n_bands_converge=$(n_bands_converge + 3)`)")
+    end
+
+    # TODO We should probably set occupation values below occupation_threshold explicitly
+    #      to zero and ignore these bands in the density computation
+    ρout = compute_density(ham.basis, eigres.X, occupation)
+    (ψ=eigres.X, eigenvalues=eigres.λ, occupation, εF, ρout, diagonalization=eigres,
+     n_bands_converge, bands.occupation_threshold)
+end
+
+
+# Band adjustment during SCF
+function default_n_bands(model)
+    n_spin = model.n_spin_components
+    min_n_bands = div(model.n_electrons, n_spin * filled_occupation(model), RoundUp)
+    n_extra = iszero(model.temperature) ? 0 : ceil(Int, 0.15 * min_n_bands)
+    min_n_bands + n_extra
+end
+default_occupation_threshold() = 1e-6
+
+"""
+In each SCF step converge exactly `n_bands_converge`, computing along the way exactly
+`n_bands_compute` (usually a few more to ease convergence in systems with small gaps).
+"""
+struct FixedBands
```

Comment:
make them a subtype of NbandsPolicy or something

Reply:
This is quite a bit to do in self_consistent_field.jl, move it to a separate file?